# 4.9 Task Part 1

# Importing libraries

In [1]:
# Import libraries
import pandas as pd
import numpy as np
import os
import matplotlib.pyplot as plt
import seaborn as sns
import scipy

# Importing Dataframe

In [2]:
# Folder path
path = '/Users/katherinevaldivia/Desktop/Career Foundry Data Analytics Course work/Data Analytics Immersion/Achievement 4/06-23-Instacart Basket Analysis /'
path

'/Users/katherinevaldivia/Desktop/Career Foundry Data Analytics Course work/Data Analytics Immersion/Achievement 4/06-23-Instacart Basket Analysis /'

In [3]:
#Importing customers.csv dataframe
customers = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'customers.csv'), index_col = False)

In [4]:
#Importing orders_products_merged_updated_2.pkl dataframe
ords_prods_group = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_merged_updated_2.pkl'))

# Task 4. Wrangle new customer dataset

In [5]:
#Checking columns listed in customers.csv dataframe for unnecessary or illogical names
customers.columns

Index(['user_id', 'First Name', 'Surnam', 'Gender', 'STATE', 'Age',
       'date_joined', 'n_dependants', 'fam_status', 'income'],
      dtype='object')

In [47]:
#Rename 'User_id' to user_id in customers.csv
customers.rename(columns = {'User_id' : 'user_id'}, inplace = True)

In [48]:
#Rename 'Surname' to Last Name in customers.csv
customers.rename(columns = {'Surnam' : 'Last Name'}, inplace = True)

In [49]:
#Rename 'STATE' to State in customers.csv
customers.rename(columns = {'STATE' : 'State'}, inplace = True)

In [50]:
#Rename 'date_joined' to Date_joined in customer.csv
customers.rename(columns = {'date_joined' : 'Date_joined'}, inplace = True)

In [51]:
#Rename 'n_dependents' to Number_of_dependents in customer.csv
customers.rename(columns = {'n_dependants' : 'Number_of_dependants'}, inplace = True)

In [52]:
#Rename 'fam_status' to Marital_status in customers.csv
customers.rename(columns = {'fam_status' : 'Marital_status'}, inplace = True)

In [53]:
# Rename 'income' to Income in customers.csv
customers.rename(columns = {'income' : 'Income'}, inplace = True)

In [54]:
#Confirming changes to column names
customers.head(1)

user_id First Name Last Name  Gender     State  Age Date_joined  \
0    26711    Deborah  Esquivel  Female  Missouri   48    1/1/2017   

   Number_of_dependants Marital_status  Income  
0                     3        married  165665

In [55]:
#Checking the size of the dataframe customers.csv
customers.shape

(206209, 10)

In [56]:
#Checking the basic information of customers.csv
customers.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 206209 entries, 0 to 206208
Data columns (total 10 columns):
 #   Column                Non-Null Count   Dtype 
---  ------                --------------   ----- 
 0   user_id               206209 non-null  int64 
 1   First Name            206209 non-null  object
 2   Last Name             206209 non-null  object
 3   Gender                206209 non-null  object
 4   State                 206209 non-null  object
 5   Age                   206209 non-null  int64 
 6   Date_joined           206209 non-null  object
 7   Number_of_dependants  206209 non-null  int64 
 8   Marital_status        206209 non-null  object
 9   Income                206209 non-null  int64 
dtypes: int64(4), object(6)
memory usage: 15.7+ MB


### Observation : All columns have a total of 206209 non-null entries except The column 'First Name' that has only 194950 non-null entries, meaning there are 11259 missing entries. To confirm this I will use value.counts function. 

In [57]:
#Double Checking and confirming First Name column for missing values as it is the only one with fewer non-null entries.
customers['First Name'].value_counts(dropna = False)

nan        11259
Marilyn     2213
Barbara     2154
Todd        2113
Jeremy      2104
           ...  
Merry        197
Eugene       197
Garry        191
Ned          186
David        186
Name: First Name, Length: 208, dtype: int64

### Observation: I was able to confirm the 11259 missing first names. However, it would be impossible to determine what their names would be and fill in those missing entries without consulting the source of information. I'll leave the blanks alone because it should not affect the statistics.

In [58]:
#Checking the descriptive statistics of customers.csv
customers.describe()

user_id            Age  Number_of_dependants         Income
count  206209.000000  206209.000000         206209.000000  206209.000000
mean   103105.000000      49.501646              1.499823   94632.852548
std     59527.555167      18.480962              1.118433   42473.786988
min         1.000000      18.000000              0.000000   25903.000000
25%     51553.000000      33.000000              0.000000   59874.000000
50%    103105.000000      49.000000              1.000000   93547.000000
75%    154657.000000      66.000000              3.000000  124244.000000
max    206209.000000      81.000000              3.000000  593901.000000

In [59]:
#Checking the first ten rows of customers.csv 
customers.head(10)

user_id First Name Last Name  Gender           State  Age Date_joined  \
0    26711    Deborah  Esquivel  Female        Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female      New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male           Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female            Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female        Maryland   26    1/1/2017   
5   133128    Cynthia     Noble  Female        Kentucky   43    1/1/2017   
6   152052      Chris    Walton    Male         Montana   20    1/1/2017   
7   168851     Joseph   Hickman    Male  South Carolina   30    1/1/2017   
8    69965     Jeremy      Vang    Male           Texas   47    1/1/2017   
9    82820      Shawn     Chung    Male        Virginia   26    1/1/2017   

   Number_of_dependants Marital_status  Income  
0                     3        married  165665  
1                     0         single   59285  
2                     2        married   99568  
3                     0         single   42049  
4                     1        married   40374  
5                     2        married   49643  
6                     0         single   61746  
7                     0         single   63712  
8                     1        married  162432  
9                     2        married   32072

# Task 5. Complete the fundamental data quality and consistency checks

In [60]:
#Checking for mixed-types in customers.csv
for col in customers.columns.tolist():
    weird = (customers[[col]].applymap(type) != customers[[col]].iloc[0].apply(type)).any(axis = 1)
    if len (customers[weird]) > 0:
        print(col)

### The First Name column is mixed because as previously stated it has 11259 null values. I can fix this by changing the data type to string.

In [61]:
#Change the datatype of First Name to a String
customers['First Name'] = customers['First Name'].astype('str')

In [62]:
#Checking for missing values again just to confirm data type change to First Name
customers.isnull().sum()

user_id                 0
First Name              0
Last Name               0
Gender                  0
State                   0
Age                     0
Date_joined             0
Number_of_dependants    0
Marital_status          0
Income                  0
dtype: int64

In [63]:
#Finding any Duplicate entries in customer.csv and assigning them to df_dups
df_dups = customers[customers.duplicated()]
df_dups

Empty DataFrame
Columns: [user_id, First Name, Last Name, Gender, State, Age, Date_joined, Number_of_dependants, Marital_status, Income]
Index: []

### Observation: Confirmed that there are NO duplicates in the customers.csv dataframe.

# Task 6. Combine your customer data with the rest of your prepared Instacart data.

In [64]:
#Checking ords_prods_group 
ords_prods_group.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 32404859 entries, 0 to 32404858
Data columns (total 24 columns):
 #   Column                 Dtype   
---  ------                 -----   
 0   order_id               int64   
 1   user_id                int64   
 2   order_number           int64   
 3   order_day_of_week      int64   
 4   order_hour_of_day      int64   
 5   days_since_last_order  float64 
 6   product_id             int64   
 7   add_to_cart_order      int64   
 8   reordered              int64   
 9   Unnamed: 0_y           int64   
 10  product_name           object  
 11  aisle_id               int64   
 12  department_id          int64   
 13  prices                 float64 
 14  _merge                 category
 15  busiest_day            object  
 16  busiest_days           object  
 17  busiest_period_of_day  object  
 18  max_order              int64   
 19  loyalty_flag           object  
 20  average_price          float64 
 21  spending_flag          object

In [65]:
#Checking the ords_prods_group df size
ords_prods_group.shape

(32404859, 24)

In [66]:
#Re-checking the customers.csv info
customers.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 206209 entries, 0 to 206208
Data columns (total 10 columns):
 #   Column                Non-Null Count   Dtype 
---  ------                --------------   ----- 
 0   user_id               206209 non-null  int64 
 1   First Name            206209 non-null  object
 2   Last Name             206209 non-null  object
 3   Gender                206209 non-null  object
 4   State                 206209 non-null  object
 5   Age                   206209 non-null  int64 
 6   Date_joined           206209 non-null  object
 7   Number_of_dependants  206209 non-null  int64 
 8   Marital_status        206209 non-null  object
 9   Income                206209 non-null  int64 
dtypes: int64(4), object(6)
memory usage: 15.7+ MB


In [67]:
#Combining the two df using the user_id column
combined = ords_prods_group.merge(customers, on = 'user_id')

In [68]:
#Checking first five rows of combined df
combined.head()

order_id  user_id  order_number  order_day_of_week  order_hour_of_day  \
0   2539329        1             1                  2                  8   
1   2398795        1             2                  3                  7   
2    473747        1             3                  3                 12   
3   2254736        1             4                  4                  7   
4    431534        1             5                  4                 15   

   days_since_last_order  product_id  add_to_cart_order  reordered  \
0                    NaN         196                  1          0   
1                   15.0         196                  1          1   
2                   21.0         196                  1          1   
3                   29.0         196                  1          1   
4                   28.0         196                  1          1   

   Unnamed: 0_y  ...   order_frequency_flag  First Name  Last Name  Gender  \
0           195  ...  Non-frequent customer       Linda     Nguyen  Female   
1           195  ...  Non-frequent customer       Linda     Nguyen  Female   
2           195  ...  Non-frequent customer       Linda     Nguyen  Female   
3           195  ...  Non-frequent customer       Linda     Nguyen  Female   
4           195  ...  Non-frequent customer       Linda     Nguyen  Female   

     State Age Date_joined Number_of_dependants  Marital_status Income  
0  Alabama  31   2/17/2019                    3         married  40423  
1  Alabama  31   2/17/2019                    3         married  40423  
2  Alabama  31   2/17/2019                    3         married  40423  
3  Alabama  31   2/17/2019                    3         married  40423  
4  Alabama  31   2/17/2019                    3         married  40423  

[5 rows x 33 columns]

In [69]:
#Checking size of combined df
combined.shape

(32404859, 33)

In [70]:
#Checking basic info of combined df to ensure information is correct
combined.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 32404859 entries, 0 to 32404858
Data columns (total 33 columns):
 #   Column                 Dtype   
---  ------                 -----   
 0   order_id               int64   
 1   user_id                int64   
 2   order_number           int64   
 3   order_day_of_week      int64   
 4   order_hour_of_day      int64   
 5   days_since_last_order  float64 
 6   product_id             int64   
 7   add_to_cart_order      int64   
 8   reordered              int64   
 9   Unnamed: 0_y           int64   
 10  product_name           object  
 11  aisle_id               int64   
 12  department_id          int64   
 13  prices                 float64 
 14  _merge                 category
 15  busiest_day            object  
 16  busiest_days           object  
 17  busiest_period_of_day  object  
 18  max_order              int64   
 19  loyalty_flag           object  
 20  average_price          float64 
 21  spending_flag          object

# Task 7. Ensure your notebook contains logical titles, section headings, and descriptive code comments.

In [71]:
# Done

# Task 8. Export this new dataframe as a pickle file so you can continue to use it in the second part of this task.

In [72]:
#Exporting combined customer data set with ords_prods_group to pickle file
combined.to_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'customer_ords_prods_merged.pkl'))

# Task 9. Save your notebook so that you can send it to your tutor for review after completing part 2.